### PR TITLE
Use RouteTemplate and TemplateMatcher

### PR DESCRIPTION
**Pull request by jbagga**
Addresses #438 

Design for using RouteTemplate, TemplateParser and TemplateMatcher

cc @rynowak

### CODE REVIEW

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs [hunk 1]
```diff
@@ -1,60 +1,93 @@
...
 
         public async Task Invoke(HttpContext httpContext)
         {
-            var dictionary = new Dictionary<string, DispatcherFeature>
+            foreach (var entry in _options.DispatcherEntryList)
             {
+                var parsedTemplate = entry.RouteTemplate;
+                var defaults = GetDefaults(parsedTemplate);
+                var templateMatcher = new TemplateMatcher(parsedTemplate, defaults);
+                var values = new RouteValueDictionary();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: This doesn't get used, it gets overwritten pretty much immediately.

Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: Just stick to `out var` at line 34

File: samples/DispatcherSample/Startup.cs [hunk 2]
```diff
@@ -13,6 +16,60 @@ public class Startup
...
+                options.DispatcherEntryList = new List<DispatcherEntry>()
+                {
+                    new DispatcherEntry
+                    {
+                        RouteTemplate = TemplateParser.Parse("{Endpoint=example}"),
+                        Endpoints = new []
+                        {
+                            new RouteValuesEndpoint("example")
+                            {
+                                RouteValueDictionary = new RouteValueDictionary(new { Endpoint = "First" }),
```
Comment (rynowak) [inline samples/DispatcherSample/Startup.cs]: This isn't a great name, suggestion `Values` or `RequiredValues`

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs [hunk 3]
```diff
@@ -1,60 +1,93 @@
...
-                },
-            };
+                    result.Add(parameter.Name, parameter.DefaultValue);
+                }
+            }
+
+            return result;
+        }
+
+        private bool MatchURLToEndpoint(RouteValueDictionary currentURLRouteValueDictionary, RouteValueDictionary endpointRouteValueDictionary)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: Names need love. 

1. `MatchURLToEndpoint` doesn't describe what's happening here. This is matching the route values to the endpoints
2. Suggested names for parameters `values`, `current values`

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs [hunk 4]
```diff
@@ -1,60 +1,93 @@
...
+        {
+            var endpointMatch = new RouteValueDictionary();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: This isn't use for anything other than maintaining a count, so we just use an int if you need that

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs [hunk 5]
```diff
@@ -1,60 +1,93 @@
...
+                    endpointMatch[key] = endpointRouteValueDictionary[key];
+                }
             }
+
+            if (endpointMatch.Count == currentURLRouteValueDictionary.Count)
+            {
+                return true;
+            }
+
+            return false;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: Suggestion on the algorithm.

Psuedocode
```
for each kvp in requiredValues
{
   if (string.IsNullOrEmpty(kvp.Value))
   {
      if (values.TryGetValue(kvp.Key, out var routeValue) && !string.IsNullOrEmpty(routeValue))
      {
          return false;
      }
   }
   else
   {
        if (!values.TryGetValue(kvp.Key, out var routeValue)  || string.Equals(kvp.Value, routeValue, OrdinalIgnoreCase))
        {
            return false;
         }
    }
}

return true;
```

Fun Facts:
1. All route values need to be converted to strings so they can be compared (usually they already are strings)
2. All route value strings are compared ignoring case
3. Null and empty string are considered equivalent
5. All of the values in the endpoint's route value collection that have a null/empty value require that the corresponding value from the route is null/empty/not-present
6. All of the values in the endpoint's route value collection that have a non-null value require that the corresponding value from the route is equal
7. The route values provided by the route can and will have more values (`.Count`) than the route values required by the endpoint

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherOptions.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+using Microsoft.AspNetCore.Routing.Template;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    public class DispatcherOptions
+    {
+        public IList<DispatcherEntry> DispatcherEntryList { get; set; }
+    }
+
+    public class DispatcherEntry
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/DispatcherOptions.cs]: This gets its own file

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs (hunk 1 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: But it isn't an out parameter

Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: OH. You're right by golly. You need to clear it then if the route matches, but none of the endpoints do. TryMatch will dirty the dictionary

**Comment by jbagga** [comment]
@rynowak Do we want to work on the route value matching algorithm as part of this PR?

File: src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs [hunk 7]
```diff
@@ -1,60 +1,97 @@
...
-                            RequestDelegate = async (context) =>
+                            var dispatcherFeature = new DispatcherFeature
                             {
-                                await context.Response.WriteAsync("Hello from the second example!");
-                            }
+                                Endpoint = endpoint,
+                                RequestDelegate = endpoint.RequestDelegate
+                            };
+
+                            httpContext.Features.Set<IDispatcherFeature>(dispatcherFeature);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs]: You also need to a `break` here. You don't want to process multiple entries - we should move on once we have a match

**Comment by rynowak** [comment]
I think it's OK if you fix the algorithm in a follow up. You need to fix this one https://github.com/aspnet/Routing/pull/443#discussion_r139012358

**Comment by jbagga** [comment]
@rynowak Added `break` to the loop

**Pull request merged**